### PR TITLE
TCP endpoint rediscovery

### PR DESCRIPTION
**Pull request by ealsur**
# Pull Request Template

## Description

This PR adds a new Direct/TCP configuration to enable TCP end-of-connection detection for proactive address refresh.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -404,6 +404,17 @@ public CosmosSerializer Serializer
...
+        /// <summary>
+        /// Gets or sets the flag to enable address cache refresh on TCP connection reset notification.
+        /// </summary>
+        /// <remarks>
+        /// Does not apply if <see cref="ConnectionMode.Gateway"/> is used.
+        /// </remarks>
+        /// <value>
+        /// The default value is false
+        /// </value>
+        public bool EnableTcpConnectionEndpointRediscovery { get; set; } = false;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Should this just be enabled by default? In what scenario does a user not want this?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Yes, gated on HA team ack.

File: Microsoft.Azure.Cosmos/src/ConnectionPolicy.cs [hunk 2]
```diff
@@ -179,6 +179,18 @@ public Protocol ConnectionProtocol
             set;
         }
 
+        /// <summary>
+        /// Gets or sets the flag to enable address cache refresh on connection reset notification.
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ConnectionPolicy.cs]: Caches etc.. might be confusing. How about calling it out as faster re-discovery of endpoitns?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ConnectionPolicy.cs]: This is not the public documentation, this is the internal ConnectionPolicy I copied from V2

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 1 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: We are going with false in V3 and true in V4.

**Pull request merged**